filters/openpolicyagent/openpolicyagent.go

```diff
@@ -516,6 +604,109 @@ func (opa *OpenPolicyAgentInstance) Start(ctx context.Context, timeout time.Dura
 	return nil
 }
 
+// StartAndTriggerPlugins Start starts the policy engine's plugin manager and triggers the plugins to download policies etc.
+func (opa *OpenPolicyAgentInstance) StartAndTriggerPlugins(ctx context.Context) error {
+	err := opa.manager.Start(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = opa.triggerPluginsWithRetry(ctx)
+	if err != nil {
+		opa.Close(ctx)
+		return err
+	}
+
+	err = opa.verifyAllPluginsStarted()
+	if err != nil {
+		opa.Close(ctx)
+		return err
+	}
+	return nil
+}
+
+func (opa *OpenPolicyAgentInstance) triggerPluginsWithRetry(ctx context.Context) error {
+	var err error
+	backoff := 100 * time.Millisecond
+	retryTrigger := time.NewTimer(backoff)
+	defer retryTrigger.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context cancelled while triggering plugins: %w, last retry returned: %w", ctx.Err(), err)
+		case <-retryTrigger.C:
+			err = opa.triggerPlugins(ctx)
+
+			if !opa.isRetryable(err) {
+				return err
+			}
+			backoff *= 2
+			retryTrigger.Reset(backoff)
+		}
+	}
+}
+
+func (opa *OpenPolicyAgentInstance) isRetryable(err error) bool {
+	var httpError download.HTTPError
+
+	if errors.As(err, &httpError) {
+		opa.Logger().WithFields(map[string]interface{}{
+			"error": httpError.Error(),
+		}).Warn("Triggering bundles failed. Response code %v, Retrying.", httpError.StatusCode)
+		return httpError.StatusCode == 429 || httpError.StatusCode >= 500
+	}
+
+	var urlError *url.Error
+	if errors.As(err, &urlError) {
+		retry := strings.Contains(urlError.Error(), "net/http: timeout awaiting response headers")
+		if retry {
+			opa.Logger().WithFields(map[string]interface{}{
+				"error": urlError.Error(),
+			}).Warn("Triggering bundles failed. Retrying.")
+		}
+		return retry
+	}
+	return false
+}
+
+func (opa *OpenPolicyAgentInstance) verifyAllPluginsStarted() error {
+	allPluginsReady := true
+	for pluginName, status := range opa.manager.PluginStatus() {
+		if status != nil && status.State != plugins.StateOK {
+			opa.Logger().WithFields(map[string]interface{}{
+				"plugin_name":   pluginName,
+				"plugin_state":  status.State,
+				"error_message": status.Message,
+			}).Error("Open policy agent plugin failed to start %s", pluginName)
+
+			allPluginsReady = false
+		}
+	}
+	if !allPluginsReady {
+		return fmt.Errorf("open policy agent plugins failed to start")
+	}
+	return nil
+}
+
+func (opa *OpenPolicyAgentInstance) triggerPlugins(ctx context.Context) error {
+
+	if opa.stopped {
```

Comment:
"stopped" is true as soon as OPA has been shut down. During the shutdown process we might still trigger the plugins. 

I wonder I we should add a new attribute "stopping" that we set before we initiate the shutdown on the plugin manager

Reply:
addressed with https://github.com/zalando/skipper/pull/3464/commits/50bacbe70463bdd58764926cc5ec4eda49934c22